### PR TITLE
Adding support to run integration tests in Jenkins

### LINKED ISSUE #263: Make all test jobs parameterized
As part of this task, we need to do -
 * Setting up right environment on jenkins for running the tests
 * Pull Job shell scripts from SCM
 * Wrap respective test.py(s) to invoke as jenkins job

### PR DESCRIPTION
**Pull request by saratvemulapalli**
### Description
Adding support to run integration tests in Jenkins.

Parameters Integration Tests need:
1. Artifacts S3 Bucket - To pull build/bundle manifests, maven dependencies.
2. OpenSearch version - To pull build/bundle manifest.
3. Build ID - To map Build ID to a test run (Test Run ID) and to pull the right manifests.
4. CPU Architecture - To identify which agent node to run on.
5. Test Run ID - To publish results for a unique job ID

### Issues Resolved
Closes #263 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (peternied) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:18]: Why are both manifest needed?

**Review by dblock — CHANGES_REQUESTED** [review]
1. It might not be possible, but a bundle defines the architecture for which it's built. This job lets me try and run a bundle for x86 on arm, which is not desirable. Can this be loaded from the bundle?
2. I'm worried that we're introducing `test_integration.sh`, which we have agreed to really be `test.sh --suite integration` or something like that. Can we do this, even if it's super hacky inside `test.sh`, in this PR so we don't have to go change it all over the place?

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (peternied) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:44]: Can we have this fail if there is no match?

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (peternied) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: Could we reshape this workflow step to be more granular and avoid having to know about s3, here is what I was thinking it would looking like:
```
steps { 
   withAWS(role: "opensearch-integration-test") {
      s3Download(file: "tests/bundle_manifest.yml", bucket: "${ARTIFACT_BUCKET_NAME}", path: "$bundle_manifest", force: true)
      sh "./bundle-workflow/download_bundle --manifest $bundle_manifest --destination /bundle_under_test"
   }
   sh "./bundle-workflow/test_integration.sh --manifest $bundle_manifest --bundle_under_test /bundle_under_test" --test-artifacts /test_artifacts-$BUNDLE_TEST_ID
}
post() {
   always {
      withAWS(role: 'opensearch-integration-test', ...) {
         sh "./bundle-workflow/collect_additional_artifacts.sh --test-artifacts /test_artifacts-${BUNDLE_TEST_ID}"
         s3Upload(file: '/test_artifacts-${architecture}-${BUNDLE_TEST_ID}', bucket: '${test_results_bucket}')
      }
   }
}
```

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (peternied) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:59]: Why do we have a `test_manifest.yml` do we need this managed as config rather than as code?

**Review by peternied — CHANGES_REQUESTED** [review]
I was expecting to see all the legs of the integration tests executed, `integration`, `bwc`, `perf` is that coming in a future PR?

**Comment by saratvemulapalli** [comment]
> I was expecting to see all the legs of the integration tests executed, `integration`, `bwc`, `perf` is that coming in a future PR?

Sure. This is a middle piece i.e bundle workflow -> test orchestrator -> individual test job. 
This helps in keeping the jobs independent while the orchestrator combines all of them into one place and executes integ tests, bwc tests and perf tests.
This PR lays down the framework of how an individual test job can run.

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:18]: As covered in the description, integration tests needs to understand what maven artifacts it needs to pull and where to pull them from for a given component. 
That said, we agreed upon having a single manifest to feed to test system which will be the source of truth.
Opened up an issue to take care of it: https://github.com/opensearch-project/opensearch-build/issues/433

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: As we discussed, eventually all the tests will consume a place to store results and pull dependencies. 
Unfortunately we have to live with it that these test systems need to understand for now.

Eventually we should be able to publish test results from the jenkins stage avoiding the test suites to manage it for them  and be agnostic of where they are being stored.

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:59]: Few tests need additional configuration explaining what kind of tests they would like to run for a given component. 
This manifest helps understand that.

**Comment by saratvemulapalli** [comment]
> 1. It might not be possible, but a bundle defines the architecture for which it's built. This job lets me try and run a bundle for x86 on arm, which is not desirable. Can this be loaded from the bundle?
> 2. I'm worried that we're introducing `test_integration.sh`, which we have agreed to really be `test.sh --suite integration` or something like that. Can we do this, even if it's super hacky inside `test.sh`, in this PR so we don't have to go change it all over the place?

1. It should be able to. Just didnt want to complicate the logic in Jenkins. But I agree and see the value that we are giving room to have mis-matched states while the truth already exists. I'll take care of this.
2. Absolutely, we went back and forth on this. I wanted to clear this up as part of #393 while not holding this one up. I'll put something really simple for this PR and we could double down on the structure with #393.

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (peternied) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: > Unfortunately we have to live with it that these test systems need to understand for now.

We can reduce a large amount of complexity in the tests if we do this.  If we are seeing trouble with the existing system please consider this an option to migrate to rather than additional investment

Comment (setiah) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: @saratvemulapalli we don't need to download manifest files from s3 here. We are doing that as part of the individual test jobs. The jenkins job should supply - s3_bucket_name, opensearch_version, build_id, architecture and test_id to the underlying tool for the test run

Comment (setiah) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: test_manifest will be part of the opensearch-build repo, so not required as input parameter

Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: Sure, the script today has it as a required parameter. 
We could always change this.

Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: Absolutely, that was the discussion we had with @peternied. Agree that this should go to orchestrator and the jobs will just expect a path to these manifests.

**Comment by peternied** [comment]
@saratvemulapalli During our discussion yesterday reading the manifest for architecture and reducing number of input parameters were mentioned - I'm waiting on those change

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (setiah) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: No, the orchestrator shouldn't be pulling these manifest files.

Since each of these jobs runs (can run) on a different host than orchestrator, the orchestrator (pipeline) shouldn't be pulling manifest files. Individual jobs should be pulling them when they run based in the input parameters - s3_bucket, opensearch_version, build_id, architecture. This logic is already added to Integ tests in #456

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (setiah) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:59]: The test manifest will be present in opensearch-build repo and doesn't need to be provided as input to the job. See #456

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: Makes sense. So we are refactoring pulling the manifests into the Python scripts and keeping the jenkins job independent of it. 
 So I would have to wait until #456 is merged. 
 
 This would mean:
1. The Jenkins job will simply read through all the parameters which integtest needs to run and these parameters could be fed from test orchestrator vs :
    a. s3 bucket
    b. opensearch version
    c. build_id
    d. test_id
    e. architecture
2. Jenkins job will not pull the manifest files as the python test suites are pulling the manifests from S3. The consequence of this is that the Jenkins job will need to pull an architecture parameter which should be passed and will not read from the manifest.

Comment (setiah) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:60]: That's correct. Instead of orchestrator, we want individual test jobs to pull the manifest and other artifacts from S3 because the jobs can run on any host irrespective of where orchestrator logic runs. This keeps it decoupled.
 
For testing purpose, you could use feature branch of 456 to unblock this.

File: bundle-workflow/README.md
Comment (dblock) [inline bundle-workflow/README.md:106]: Which manifest?

**Review by dblock — APPROVED** [review]

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (dblock) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:63]: Shouldn't this have "integ-test" in the arguments?

**Review by dblock — CHANGES_REQUESTED** [review]

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile]: It should :), thanks for that. I have updated the runner later.

File: bundle-workflow/README.md
Comment (saratvemulapalli) [reply bundle-workflow/README.md:106]: Not sure if the question was what manifest do we use for integ tests. 
I would like to have #456 update this readme to add the rest of the new parameters to this readme.

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (setiah) [inline bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:16]: This can be read from environment right? Does it need to be a parameter?

File: bundle-workflow/test.sh
Comment (setiah) [inline bundle-workflow/test.sh:12]: This change is also present in #456 now as requested. Can we remove it from here?

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:16]: Yeah, the idea is test orchestrator tells which bucket to run with. And jobs are agnostic of the ENV even though they have access to it. This helps when a user wants to submit a job and use a different S3 bucket.

File: bundle-workflow/test.sh
Comment (saratvemulapalli) [reply bundle-workflow/test.sh:12]: Sure that works :)

File: bundle-workflow/jenkins_workflow/integtest/Jenkinsfile
Comment (setiah) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:16]: Changing bucket may not work because the jenkins role won't have permissions to read from a different s3 bucket. The role is also configured in jenkins env variables.

Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/Jenkinsfile:16]: Thinking about it. This actually makes sense. May be lets not optimize for other use cases let one flow work concretely. I'll use the ENV variable.

**Review by setiah — DISMISSED** [review]
LGTM. Thanks @saratvemulapalli

File: bundle-workflow/test.sh
Comment (dblock) [inline bundle-workflow/test.sh:12]: Did you delete src/test.py?

File: bundle-workflow/jenkins_workflow/integtest/README.md
Comment (dblock) [inline bundle-workflow/jenkins_workflow/integtest/README.md:5]: I don't think this is correct anymore.

File: bundle-workflow/test.sh
Comment (saratvemulapalli) [reply bundle-workflow/test.sh:12]: Yes, @setiah will add the test.sh in #456 as part of his PR. 
Ref: https://github.com/opensearch-project/opensearch-build/pull/423#discussion_r708765944

File: bundle-workflow/jenkins_workflow/integtest/README.md
Comment (saratvemulapalli) [reply bundle-workflow/jenkins_workflow/integtest/README.md:5]: You are right. Let me fix it.

**Review by dblock — APPROVED** [review]

**Review by setiah — APPROVED** [review]

**Pull request merged**
